src/Storages/Statistics/ConditionSelectivityEstimator.cpp

```diff
@@ -135,38 +171,201 @@ void ConditionSelectivityEstimator::ColumnSelectivityEstimator::addStatistics(St
+            }
+        },
+        {
+            "in",
+            [] (RPNElement & out, const String & column, const Field & value)
+            {
+                out.function = RPNElement::FUNCTION_IN_RANGE;
+                Ranges ranges;
+                for (const Field & field : value.safeGet<Tuple>())
+                {
+                    ranges.emplace_back(field);
+                }
+                out.column_ranges.emplace(column, PlainRanges(ranges, /*intersect*/ true, /*ordered*/ false));
+            }
+        },
+        {
+            "notIn",
+            [] (RPNElement & out, const String & column, const Field & value)
+            {
+                out.function = RPNElement::FUNCTION_IN_RANGE;
+                Ranges ranges;
+                for (const Field & field : value.safeGet<Tuple>())
+                {
+                    ranges.emplace_back(field);
+                }
+                out.column_not_ranges.emplace(column, PlainRanges(ranges, /*intersect*/ true, /*ordered*/ false));
+            }
+        },
+        {
+            "less",
+            [] (RPNElement & out, const String & column, const Field & value)
+            {
+                out.function = RPNElement::FUNCTION_IN_RANGE;
+                out.column_ranges.emplace(column, Range::createRightBounded(value, false));
+            }
+        },
+        {
+            "greater",
+            [] (RPNElement & out, const String & column, const Field & value)
+            {
+                out.function = RPNElement::FUNCTION_IN_RANGE;
+                out.column_ranges.emplace(column, Range::createLeftBounded(value, false));
+            }
+        },
+        {
+            "lessOrEquals",
+            [] (RPNElement & out, const String & column, const Field & value)
+            {
+                out.function = RPNElement::FUNCTION_IN_RANGE;
+                out.column_ranges.emplace(column, Range::createRightBounded(value, true));
+            }
+        },
+        {
+            "greaterOrEquals",
+            [] (RPNElement & out, const String & column, const Field & value)
+            {
+                out.function = RPNElement::FUNCTION_IN_RANGE;
+                out.column_ranges.emplace(column, Range::createLeftBounded(value, true));
+            }
+        }
+};
+
+/// merge CNF or DNF
+bool ConditionSelectivityEstimator::RPNElement::tryToMergeClauses(RPNElement & lhs, RPNElement & rhs)
 {
-    return rows - estimateLess(val, rows);
+    auto canMergeWith = [&](const RPNElement & e)
+    {
+        return (e.function == FUNCTION_IN_RANGE
+                /// if the sub-clause is also cnf/dnf, it's good to merge
+                || e.function == function
+                /// if the sub-clause is different, but has only one column, it also works, e.g
+                /// (a > 0 and a < 5) or (a > 3 and a < 10) can be merged to (a > 0 and a < 10)
+                || e.column_ranges.size() + e.column_not_ranges.size() == 1
```

Comment:
```suggestion
                || (e.column_ranges.size() + e.column_not_ranges.size()) == 1
```
Looks safer and easier to read